libs/metrics-wai/src/Data/Metrics/Types.hs

```diff
@@ -0,0 +1,62 @@
+{-# LANGUAGE FlexibleContexts    #-}
+{-# LANGUAGE OverloadedStrings   #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeFamilies        #-}
+{-# LANGUAGE ViewPatterns        #-}
+
+module Data.Metrics.Types
+    ( PathTemplate(..)
+    , Paths(..)
+    , mkTree
+    , treeLookup
+    ) where
+
+import Imports
+import Data.Tree as Tree
+
+import qualified Data.ByteString.Char8 as BS
+
+
+newtype PathTemplate = PathTemplate Text
+
+-- | A 'Forest' of path segments.  A path segment is 'Nothing' if it captures a value
+-- (e.g. user id).
+newtype Paths = Paths (Forest (Maybe ByteString))
+  deriving (Eq, Show)
+
+-- | Turn a list of paths into a 'Paths' tree.  Treat all path segments that start with @':'@
+-- as equal and turn them into 'Nothing'.
+mkTree :: forall m. (m ~ Either String) => [[ByteString]] -> m Paths
```

Comment:
Just for fun I built this with Cofree and a custom datatype; remind me tomorrow and I'll walk you through Cofree (just for fun 😄 )

It's probably a bit confusing to read (maybe best re-written with a simple recursive data-type rather than Cofree); but I believe it fixes all the bugs we discovered yesterday evening. That said, it looks like you took another pass at this one, so if yours is working, probably best to stick with it 👍  mine is unnecessarily complicated.

https://gist.github.com/ChrisPenner/829279ef928512b84ce1448241f90aa8